File: apps/react-native/src/data/mappers.ts
// Defensive DTO→domain mapping. Every function validates raw JSON and throws MappingError
// on anything untrusted: a missing field, a wrong type, an unknown enum, an unparseable
// date, or a domain-invariant violation. The executor turns that into MalformedResponse.

import {
  Event,
  EventDetail,
  EventPage,
  EventStatus,
  money,
  Order,
  OrderStatus,
  QueueStatus,
  QueueToken,
  Reservation,
  ReservationStatus,
  Sector,
  sector,
  TokenPair,
} from '../domain/models';
import type { EventDto, OrderDto, QueueTokenDto, ReservationDto } from '../contract/dto';

export class MappingError extends Error {}

type Json = Record<string, unknown>;

function asObject(json: unknown): Json {
  if (typeof json === 'object' && json !== null && !Array.isArray(json)) return json as Json;
  throw new MappingError('expected a JSON object');
}

function str(json: Json, key: string): string {
  const v = json[key];
  if (typeof v === 'string') return v;
  throw new MappingError(`missing or non-string field '${key}'`);
}

function int(json: Json, key: string): number {
  const v = json[key];
  if (typeof v === 'number' && Number.isInteger(v)) return v;
  throw new MappingError(`missing or non-integer field '${key}'`);
}

function strOrNull(json: Json, key: string): string | null {
  const v = json[key];
  if (v === null || v === undefined) return null;
  if (typeof v === 'string') return v;
  throw new MappingError(`non-string field '${key}'`);
}

function date(json: Json, key: string): Date {
  const raw = str(json, key);
  const d = new Date(raw);
  if (Number.isNaN(d.getTime())) throw new MappingError(`unparseable timestamp '${raw}' in '${key}'`);
  return d;
}

function guarded<T>(build: () => T): T {
  try {
    return build();
  } catch (e) {
    if (e instanceof MappingError) throw e;
    throw new MappingError(e instanceof Error ? e.message : 'domain invariant violated');
  }
}

// The enum maps are typed against the GENERATED contract unions (EventDto['status'], …). If
// the contract ever adds or renames a status, these object literals stop being exhaustive and
// the app fails to compile until the new case is handled. That is the payoff of the codegen.
const EVENT_STATUS: Record<EventDto['status'], EventStatus> = { draft: 'draft', on_sale: 'onSale', sold_out: 'soldOut', closed: 'closed' };
const QUEUE_STATUS: Record<QueueTokenDto['status'], QueueStatus> = { waiting: 'waiting', admitted: 'admitted', expired: 'expired' };
const RESERVATION_STATUS: Record<ReservationDto['status'], ReservationStatus> = { held: 'held', confirmed: 'confirmed', released: 'released', expired: 'expired' };
const ORDER_STATUS: Record<OrderDto['status'], OrderStatus> = { pending: 'pending', paid: 'paid', failed: 'failed', refunded: 'refunded' };

function mapEnum<K extends string, T>(table: Record<K, T>, raw: string, what: string): T {
  if (raw in table) return table[raw as K];
  throw new MappingError(`unknown ${what} '${raw}'`);
}

export function eventFromJson(json: unknown): Event {
  const m = asObject(json);
  return {
    id: str(m, 'id'),
    name: str(m, 'name'),
    venue: str(m, 'venue'),
    startsAt: date(m, 'starts_at'),
    salesOpenAt: date(m, 'sales_open_at'),
    status: mapEnum(EVENT_STATUS, str(m, 'status'), 'event status'),
  };
}

export function sectorFromJson(json: unknown): Sector {
  const m = asObject(json);
  return guarded(() =>
    sector({
      id: str(m, 'id'),
      eventId: str(m, 'event_id'),
      name: str(m, 'name'),
      price: money(int(m, 'price_cents'), str(m, 'currency')),
      totalInventory: int(m, 'total_inventory'),
      availableInventory: int(m, 'available_inventory'),
    }),
  );
}

export function eventDetailFromJson(json: unknown): EventDetail {
  const m = asObject(json);
  const sectors = m['sectors'];
  if (!Array.isArray(sectors)) throw new MappingError('missing sectors');
  return { event: eventFromJson(m), sectors: sectors.map(sectorFromJson) };
}

export function eventPageFromJson(json: unknown): EventPage {
  const m = asObject(json);
  const data = m['data'];
  if (!Array.isArray(data)) throw new MappingError('missing data');
  return { events: data.map(eventFromJson), nextCursor: strOrNull(m, 'next_cursor') };
}

export function queueTokenFromJson(json: unknown): QueueToken {
  const m = asObject(json);
  return {
    id: str(m, 'id'),
    userId: str(m, 'user_id'),
    eventId: str(m, 'event_id'),
    position: int(m, 'position'),
    status: mapEnum(QUEUE_STATUS, str(m, 'status'), 'queue status'),
    admittedAt: m['admitted_at'] == null ? null : date(m, 'admitted_at'),
  };
}

export function reservationFromJson(json: unknown): Reservation {
  const m = asObject(json);
  return guarded(() => ({
    id: str(m, 'id'),
    userId: str(m, 'user_id'),
    sectorId: str(m, 'sector_id'),
    quantity: int(m, 'quantity'),
    status: mapEnum(RESERVATION_STATUS, str(m, 'status'), 'reservation status'),
    expiresAt: date(m, 'expires_at'),
  }));
}

export function orderFromJson(json: unknown): Order {
  const m = asObject(json);
  return {
    id: str(m, 'id'),
    reservationId: str(m, 'reservation_id'),
    userId: str(m, 'user_id'),
    amountCents: int(m, 'amount_cents'),
    status: mapEnum(ORDER_STATUS, str(m, 'status'), 'order status'),
    createdAt: date(m, 'created_at'),
  };
}

export function tokenPairFromJson(json: unknown): TokenPair {
  const m = asObject(json);
  return {
    accessToken: str(m, 'access_token'),
    refreshToken: str(m, 'refresh_token'),
    expiresInSeconds: int(m, 'expires_in'),
  };
}

File: apps/flutter/lib/data/mappers.dart
/// Defensive DTO→domain mapping. Every function validates the raw JSON and throws
/// [MappingException] on anything it cannot trust: a missing field, a wrong type, an
/// unknown enum value, an unparseable date, or a domain-invariant violation. The executor
/// turns that into a [MalformedResponse]. This is the zero-trust boundary.
library;

import '../domain/models.dart';

/// Thrown when a payload cannot be turned into a valid domain model.
class MappingException implements Exception {
  final String message;
  MappingException(this.message);
  @override
  String toString() => 'MappingException: $message';
}

String _str(Map json, String key) {
  final v = json[key];
  if (v is String) return v;
  throw MappingException("missing or non-string field '$key'");
}

int _int(Map json, String key) {
  final v = json[key];
  if (v is int) return v;
  throw MappingException("missing or non-int field '$key'");
}

String? _strOrNull(Map json, String key) {
  final v = json[key];
  if (v == null) return null;
  if (v is String) return v;
  throw MappingException("non-string field '$key'");
}

DateTime _date(Map json, String key) {
  final raw = _str(json, key);
  try {
    return DateTime.parse(raw);
  } on FormatException {
    throw MappingException("unparseable timestamp '$raw' in '$key'");
  }
}

Map _asMap(dynamic json) {
  if (json is Map) return json;
  throw MappingException('expected a JSON object');
}

EventStatus _eventStatus(String raw) => switch (raw) {
      'draft' => EventStatus.draft,
      'on_sale' => EventStatus.onSale,
      'sold_out' => EventStatus.soldOut,
      'closed' => EventStatus.closed,
      _ => throw MappingException("unknown event status '$raw'"),
    };

QueueStatus _queueStatus(String raw) => switch (raw) {
      'waiting' => QueueStatus.waiting,
      'admitted' => QueueStatus.admitted,
      'expired' => QueueStatus.expired,
      _ => throw MappingException("unknown queue status '$raw'"),
    };

ReservationStatus _reservationStatus(String raw) => switch (raw) {
      'held' => ReservationStatus.held,
      'confirmed' => ReservationStatus.confirmed,
      'released' => ReservationStatus.released,
      'expired' => ReservationStatus.expired,
      _ => throw MappingException("unknown reservation status '$raw'"),
    };

OrderStatus _orderStatus(String raw) => switch (raw) {
      'pending' => OrderStatus.pending,
      'paid' => OrderStatus.paid,
      'failed' => OrderStatus.failed,
      'refunded' => OrderStatus.refunded,
      _ => throw MappingException("unknown order status '$raw'"),
    };

/// Wraps domain-invariant [ArgumentError]s as [MappingException].
T _guarded<T>(T Function() build) {
  try {
    return build();
  } on ArgumentError catch (e) {
    throw MappingException(e.message?.toString() ?? 'domain invariant violated');
  }
}

Event eventFromJson(dynamic json) {
  final m = _asMap(json);
  return _guarded(() => Event(
        id: _str(m, 'id'),
        name: _str(m, 'name'),
        venue: _str(m, 'venue'),
        startsAt: _date(m, 'starts_at'),
        salesOpenAt: _date(m, 'sales_open_at'),
        status: _eventStatus(_str(m, 'status')),
      ));
}

Sector sectorFromJson(dynamic json) {
  final m = _asMap(json);
  return _guarded(() => Sector(
        id: _str(m, 'id'),
        eventId: _str(m, 'event_id'),
        name: _str(m, 'name'),
        price: Money(_int(m, 'price_cents'), _str(m, 'currency')),
        totalInventory: _int(m, 'total_inventory'),
        availableInventory: _int(m, 'available_inventory'),
      ));
}

EventDetail eventDetailFromJson(dynamic json) {
  final m = _asMap(json);
  final sectors = (m['sectors'] as List?) ?? (throw MappingException('missing sectors'));
  return EventDetail(event: eventFromJson(m), sectors: sectors.map(sectorFromJson).toList());
}

EventPage eventPageFromJson(dynamic json) {
  final m = _asMap(json);
  final data = (m['data'] as List?) ?? (throw MappingException('missing data'));
  return EventPage(events: data.map(eventFromJson).toList(), nextCursor: _strOrNull(m, 'next_cursor'));
}

QueueToken queueTokenFromJson(dynamic json) {
  final m = _asMap(json);
  return _guarded(() => QueueToken(
        id: _str(m, 'id'),
        userId: _str(m, 'user_id'),
        eventId: _str(m, 'event_id'),
        position: _int(m, 'position'),
        status: _queueStatus(_str(m, 'status')),
        admittedAt: m['admitted_at'] == null ? null : _date(m, 'admitted_at'),
      ));
}

Reservation reservationFromJson(dynamic json) {
  final m = _asMap(json);
  return _guarded(() => Reservation(
        id: _str(m, 'id'),
        userId: _str(m, 'user_id'),
        sectorId: _str(m, 'sector_id'),
        quantity: _int(m, 'quantity'),
        status: _reservationStatus(_str(m, 'status')),
        expiresAt: _date(m, 'expires_at'),
      ));
}

Order orderFromJson(dynamic json) {
  final m = _asMap(json);
  return Order(
    id: _str(m, 'id'),
    reservationId: _str(m, 'reservation_id'),
    userId: _str(m, 'user_id'),
    amountCents: _int(m, 'amount_cents'),
    status: _orderStatus(_str(m, 'status')),
    createdAt: _date(m, 'created_at'),
  );
}
